databricks/koalas/frame.py

```diff
@@ -7830,43 +7830,89 @@ def reindex(
+            kser = ks.Series(list(index))
+            labels = kser._internal.spark_frame.select(kser.spark.column.alias(index_columns[0]))
+            index_map = self._internal._index_map
+
         if fill_value is not None:
-            frame_index_column = verify_temp_column_name(frame, "__frame_index_column__")
-            frame = frame.withColumnRenamed(index_column, frame_index_column)
+            frame_index_columns = [
+                verify_temp_column_name(frame, "__frame_index_column_{}__".format(i))
+                for i in range(nlevels)
+            ]
+            index_scols = [
+                scol_for(frame, index_col).alias(frame_index_col)
+                for index_col, frame_index_col in zip(index_columns, frame_index_columns)
+            ]
+            scols = self._internal.resolved_copy.data_spark_columns
+            frame = frame.select(index_scols + scols)
 
             temp_fill_value = verify_temp_column_name(frame, "__fill_value__")
             labels = labels.withColumn(temp_fill_value, F.lit(fill_value))
 
-            frame_index_scol = scol_for(frame, frame_index_column)
-            labels_index_scol = scol_for(labels, index_column)
+            frame_index_scols = [scol_for(frame, col) for col in frame_index_columns]
+            labels_index_scols = [scol_for(labels, col) for col in index_columns]
+
+            joined_df = frame.join(
+                labels,
+                on=[fcol == lcol for fcol, lcol in zip(frame_index_scols, labels_index_scols)],
+                how="right",
+            )
 
-            joined_df = frame.join(labels, on=[frame_index_scol == labels_index_scol], how="right")
             joined_df = joined_df.select(
-                labels_index_scol,
+                *labels_index_scols,
                 *[
                     F.when(
-                        frame_index_scol.isNull() & labels_index_scol.isNotNull(),
+                        reduce(
+                            lambda c1, c2: c1 & c2,
+                            [
+                                fcol.isNull() & lcol.isNotNull()
+                                for fcol, lcol in zip(frame_index_scols, labels_index_scols)
+                            ],
+                        ),
                         scol_for(joined_df, temp_fill_value),
                     )
                     .otherwise(scol_for(joined_df, col))
                     .alias(col)
                     for col in self._internal.data_spark_column_names
-                ]
+                ],
             )
         else:
-            joined_df = frame.join(labels, on=index_column, how="right")
+            joined_df = frame.join(labels, on=index_columns, how="right")
 
-        internal = self._internal.with_new_sdf(joined_df)
+        sdf = joined_df.drop(NATURAL_ORDER_COLUMN_NAME)
+        internal = InternalFrame(
+            sdf,
+            index_map=index_map,
+            column_labels=self._internal._column_labels,
```

Comment:
nit: `self._internal.column_labels`? We'd prefer to use the property without starting with `_`.